File: leetcodeproblem/4sum.py
class Solution:
    def fourSum(self, nums, target: int):  
        nums.sort()
        res = [] 
        n = len(nums)
        for i in range(n-3):
            if i and nums[i] == nums[i-1] :
               continue 
            for j in range(i+1,n-2):
                if j > i+1 and nums[j] == nums[j-1] :
                    continue 

                l = j+1
                r = n-1
                while l < r:
                    sum = nums[i] +nums[j] +nums[l] +nums[r] 
                    if target == sum :
                       res.append([nums[i],nums[j],nums[l] ,nums[r]])
                       l+=1
                       r-=1
                       while l <r and  nums[l] == nums[l-1]:
                            l+=1

                       while l  <  r and nums[r]  == nums[r+1]:
                            r-=1
                    elif target>sum :
                         l+=1
                    else:
                        r-=1 
        return res

File: leetcodeproblem/array/4sum.js
var fourSum = function(nums, target) {
    // sorted
    //  
      nums.sort((a,b)=>a-b) 
    let res =  [] 
     for(let i=0; i< nums.length -3; i++ ){

          
             if(i > 0 &&  nums[i-1] == nums[i]){
                  continue
             }

             for(let j=i+1; j < nums.length-2 ; j++){
                    if(j > i+1 &&  nums[j-1] == nums[j]){
                       continue
                    }
                   let left = j+1 
                   let right = nums.length -1
                   while(left < right){
                        let sum =  nums[i] + nums[j] + nums[left] + nums[right];
                        // console.log(sum,[nums[i] , nums[j] , nums[left], nums[right]]);
                        
                        if(sum === target){
                                res.push([nums[i] , nums[j] , nums[left], nums[right]])
                                while (nums[left +  1]  == nums[left]) {
                                      left++
                                }
                                while (nums[right -  1]  == nums[right]) {
                                      right--
                                }
                                left++
                                right --
                        }else if(sum  < target){
                                left++
                        }else{
                             right--
                        }


                   }

               }
    }
    return res
};

console.log(fourSum([2,2,2,2,2,2],8));
